### PR TITLE
[DRAFT] Support for Prepare/Execute statement in Hive

### PR DESCRIPTION
**Pull request by vineetgarg02**
## NOTICE

Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HIVE-XXXXX: Fix a typo in YYY)
For more details, please see https://cwiki.apache.org/confluence/display/Hive/HowToContribute

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    return updatedExprList;
+  }
+
+  // Note about type inference
+  // Since dynamic parameter lacks type we need to figure out appropriate type to create constant
+  // out of string value. To do this, we choose the type of first child of the parent expression
+  // which isn't dynamic parameter
+  // TODO: cases to consider/cover
+  //  exprs have no children, expres have all dynamic parameters
+  private static ExprNodeDesc replaceDynamicParamsWithConstant(ExprNodeDesc expr,
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java:160]: This is a difficult problem indeed. I guess this assumption is fine right now but it seems it can cause different behavior compared to not using prepare statement. We should be clear about this and discuss how we will do inference moving forward (we should check the standard to see how they deal with it). In addition, I am wondering whether, for the time being, we should limit dynamic parameters to filter expressions because of this (it seems replacing for project may have more side effects over the plan).

File: ql/src/java/org/apache/hadoop/hive/ql/QueryPlan.java [hunk 2]
```diff
@@ -106,10 +98,12 @@
   private org.apache.hadoop.hive.ql.plan.api.Query query;
   private final Map<String, Map<String, Long>> counters =
       new ConcurrentHashMap<String, Map<String, Long>>();
-  private final Set<String> done = Collections.newSetFromMap(new
-      ConcurrentHashMap<String, Boolean>());
-  private final Set<String> started = Collections.newSetFromMap(new
-      ConcurrentHashMap<String, Boolean>());
+  //private final Set<String> done = Collections.newSetFromMap(new
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/QueryPlan.java:101]: I did not understand this change?

File: ql/src/java/org/apache/hadoop/hive/ql/Compiler.java [hunk 3]
```diff
@@ -329,7 +357,34 @@ private void recordValidTxns(HiveTxnManager txnMgr) throws LockException {
...
+
+      SessionState ss = SessionState.get();
+      if (ss != null) {
+        if (ss.getPreparePlans().containsKey(queryName)) {
+          QueryPlan plan = ss.getPreparePlans().get(queryName);
+          QueryPlan planCopy = PreparePlanUtils.makeCopy(plan);
+          bindDynamicParams(planCopy, getParameterMap());
+          // This is no longer prepare plan
+          planCopy.setIsPrepareQuery(false);
+          //TODO: this will create empty schema, need to figure out a way to create proper schema
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/Compiler.java:373]: Once we have retrieved the plan at this stage, we probably need to change query id to avoid any issue, i.e., multiple queries with same query id? Also we could update query text in plan and I am not sure if anything else is needed.

File: ql/src/java/org/apache/hadoop/hive/ql/Compiler.java [hunk 4]
```diff
@@ -72,19 +65,52 @@
...
   public Compiler(Context context, DriverContext driverContext, DriverState driverState) {
     this.context = context;
     this.driverContext = driverContext;
     this.driverState = driverState;
   }
 
+  /*
+   * Give string command determine the type of statement
+   */
+  private StatementType getStatementType(final ASTNode tree) {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/Compiler.java:95]: I was hoping this orchestration could be moved to the different analyzers for `PREPARE` and `EXECUTE`, e.g., similar to `ExplainSemanticAnalyzer` or others. It may make the code easier to understand instead of creating different workflows in the compiler depending on the statement type (I think `Compiler` is mostly agnostic from statement type and would probably be good to try to keep it that way).

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTBuilder.java [hunk 5]
```diff
@@ -375,6 +375,15 @@ public static ASTNode literal(RexLiteral literal) {
     return (ASTNode) ParseDriver.adaptor.create(type, String.valueOf(val));
   }
 
+  public static ASTNode dynamicParam(RexDynamicParam param) {
+    ASTNode node = (ASTNode)ParseDriver.adaptor.create(HiveParser.TOK_PARAMETER,
+        Integer.toString(param.getIndex()));
+    //ASTNode child = (ASTNode)ParseDriver.adaptor.create(HiveParser.TOK_PARAMETER_IDX,
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTBuilder.java:381]: nit. Commented code

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/PrepareStatementAnalyzer.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.hadoop.hive.ql.parse.*;
+
+import java.util.*;
+
+/**
+ * Analyzer for table dropping commands.
+ */
+@DDLType(types = HiveParser.TOK_PREPARE)
+public class PrepareStatementAnalyzer extends CalcitePlanner {
+  /*public DropTableAnalyzer(QueryState queryState) throws SemanticException {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/PrepareStatementAnalyzer.java:39]: nit. Commented code

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/ExecuteStatementAnalyzer.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+
+import java.util.*;
+
+/**
+ * Analyzer for table dropping commands.
+ */
+@DDLType(types = HiveParser.TOK_EXECUTE)
+public class ExecuteStatementAnalyzer extends BaseSemanticAnalyzer {
+  /*public DropTableAnalyzer(QueryState queryState) throws SemanticException {
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/ExecuteStatementAnalyzer.java:43]: nit. Commented code

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/ExecuteStatementAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.hadoop.hive.ql.parse.ASTNode;
+import org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer;
+import org.apache.hadoop.hive.ql.parse.HiveParser;
+import org.apache.hadoop.hive.ql.parse.ReplicationSpec;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+
+import java.util.*;
+
+/**
+ * Analyzer for table dropping commands.
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/ExecuteStatementAnalyzer.java:39]: Update comment

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/PrepareStatementAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.hadoop.hive.ql.exec.TaskFactory;
+import org.apache.hadoop.hive.ql.hooks.ReadEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.parse.*;
+
+import java.util.*;
+
+/**
+ * Analyzer for table dropping commands.
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/drop/PrepareStatementAnalyzer.java:35]: Update comment

File: ql/src/java/org/apache/hadoop/hive/ql/plan/ExprDynamicParamDesc.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+
+/**
+ * A constant expression.
+ */
+public class ExprDynamicParamDesc extends ExprNodeDesc implements Serializable {
+  private static final long serialVersionUID = 1L;
+  final protected transient static char[] hexArray = "0123456789ABCDEF".toCharArray();
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/ExprDynamicParamDesc.java:41]: Not used?

File: ql/src/test/queries/clientpositive/prepare_plan.q [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+-- same query, different param
+execute p1 (1, '0');
+select * from src where key > '0' limit 10;
+
+-- different param type
+execute p1 (1,0);
+select * from src where key > 0 limit 10;
+
+-- multiple parameters
+explain prepare p2 as select min(ctinyint), max(cbigint) from alltypesorc where cint > ($1 + $2 + $3) group by ctinyint;
```
Comment (jcamachor) [inline ql/src/test/queries/clientpositive/prepare_plan.q:27]: Cool! :)

File: ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+  protected static String getPrepareStatementName(ASTNode tree) {
+    if (tree.getType() == HiveParser.TOK_EXPLAIN) {
+      tree = (ASTNode)tree.getChildren().get(0);
+    }
+    assert (tree.getType() == HiveParser.TOK_PREPARE
+        || tree.getType() == HiveParser.TOK_EXECUTE);
+    return ((ASTNode)tree.getChildren().get(1)).getText();
+  }
+
+  public static void bindDynamicParams(QueryPlan plan, Map<Integer, String> parameterMap) {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java:86]: In order to avoid modifying the plan and adding per operator specific code below maybe we could use a UDF when generating the plan with dynamic parameters that performs a lookup in the map to obtain the correct value. Maybe other parts of the code could also be simplified like this.

File: ql/src/java/org/apache/hadoop/hive/ql/Compiler.java [hunk 13]
```diff
@@ -72,19 +65,52 @@
...
+  protected final DriverContext driverContext;
   private final DriverState driverState;
   private final PerfLogger perfLogger = SessionState.getPerfLogger();
 
   private ASTNode tree;
 
+  /*
+   * Compiler needs to know if given statement is PREPARE/EXECUTE or otherwise
+   * to orchestrate the planning.
+   * e.g. if plan should be stored or if parameter binding should be done.
```
[the remark this answers is not in the document]
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/Compiler.java:78]: I am thinking something in the lines of Postgres plan_cache_mode but maybe less ambitious to begin with.

_By default (that is, when plan_cache_mode is set to auto), the server will automatically choose whether to use a generic or custom plan for a prepared statement that has parameters. The current rule for this is that the first five executions are done with custom plans and the average estimated cost of those plans is calculated. Then a generic plan is created and its estimated cost is compared to the average custom-plan cost. Subsequent executions use the generic plan if its cost is not so much higher than the average custom-plan cost as to make repeated replanning seem preferable._

In other words use the SQL string or something equivalent as the key for the cache.

File: ql/src/test/queries/clientpositive/prepare_plan.q [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+execute pcount (1,200);
+select count(*) from src where key > '200';
+
+-- single param
+explain extended prepare p1 as select * from src where key > $1 order by key limit 10;
+explain extended select * from src where key > '200' order by key limit 10;
+prepare p1 as select * from src where key > $1 order by key limit 10;
+
+execute p1 (1,200);
+select * from src where key > '200' order by key limit 10;
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/prepare_plan.q:20]: I don't think it is necessary to execute the same query twice if it is just to show that the results are the same. If the results are verified when .out is committed I think it's sufficient.

File: ql/src/test/queries/clientpositive/prepare_plan.q [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+-- same query, different param
+execute p1 (1,0);
+select * from src where key > '0' order by key limit 10;
+
+-- multiple parameters
+explain prepare p2 as select min(ctinyint), max(cbigint) from alltypesorc where cint > ($1 + $2 + $3) group by ctinyint;
+prepare p2 as select min(ctinyint), max(cbigint) from alltypesorc where cint > ($1 + $2 + $3) group by ctinyint;
+
+execute p2 ( 1,0, 2,1, 3, 2);
+select min(ctinyint), max(cbigint) from alltypesorc where cint > (0 + 1 + 2) group by ctinyint;
```
Comment (zabetak) [inline ql/src/test/queries/clientpositive/prepare_plan.q:31]: Since we are executing the query and verifying the results we may want to add an ORDER BY clause to guarantee that the order is not gonna change.

File: parser/src/java/org/apache/hadoop/hive/ql/parse/HiveLexerParent.g [hunk 16]
```diff
@@ -471,6 +472,11 @@ Number
     (Digit)+ ( DOT (Digit)* (Exponent)? | Exponent)?
     ;
 
+ParamLiteral
+    :
+    (DOLLAR) (Digit)+ { setText(getText().substring(1, getText().length())); }
```
[the remark this answers is not in the document]
Comment (zabetak) [reply parser/src/java/org/apache/hadoop/hive/ql/parse/HiveLexerParent.g:477]: DBMS use various notations for dynamic parameters when it comes to PREPARE/EXECUTE syntax but at JDBC level I think everybody accept `?`. The SQL standard and JDBC specification use `?` for parameters so to keep things uniform I think it makes sense to also use this notation.

**Comment by vineetgarg02** [comment]
Closing this one as I had to open another one at https://github.com/apache/hive/pull/1315
